numba/cuda/simulator/vector_types.py

```diff
@@ -0,0 +1,50 @@
+from numba import types
+from numba.cuda.stubs import _vector_type_stubs
+
+
+class SimulatedVectorType:
+    attributes = ['x', 'y', 'z', 'w']
+
+    def __init__(self, *args):
+        args_flattened = []
+        for arg in args:
+            if isinstance(arg, SimulatedVectorType):
+                args_flattened += arg.as_list()
+            else:
+                args_flattened.append(arg)
+        self._attrs = self.attributes[:len(args_flattened)]
+        assert len(self._attrs) == len(args_flattened), \
```

Comment:
Ah, that's right. I meant to test that `num_elements` should be the same as `len(args_flattened)`. Since the number of elements is an intrinsic property of the vector type and the number of argument should match it. Updating.

Reply:
Done. With 7deb8dbbd we are actually testing the number of arguments matches with the number of element:
```python
In [11]: float32x3(1.0, 2.0, 3.0)
Out[11]: <numba.cuda.simulator.vector_types.float32x3 at 0x7f15810b5910>

In [12]: float32x3(1.0, 2.0, 3.0, 4.0)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [12], in <cell line: 1>()
----> 1 float32x3(1.0, 2.0, 3.0, 4.0)

File /raid/wangm/dev/numba/numba/cuda/simulator/vector_types.py:16, in SimulatedVectorType.__init__(self, *args)
     14         args_flattened.append(arg)
     15 self._attrs = self.attributes[:len(args_flattened)]
---> 16 assert self.num_elements == len(args_flattened), \
     17     f"{self.num_elements} != {len(args_flattened)}"
     19 for arg, attr in zip(args_flattened, self._attrs):
     20     setattr(self, attr, arg)

AssertionError: 3 != 4

In [13]: float32x3(1.0, 2.0)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [13], in <cell line: 1>()
----> 1 float32x3(1.0, 2.0)

File /raid/wangm/dev/numba/numba/cuda/simulator/vector_types.py:16, in SimulatedVectorType.__init__(self, *args)
     14         args_flattened.append(arg)
     15 self._attrs = self.attributes[:len(args_flattened)]
---> 16 assert self.num_elements == len(args_flattened), \
     17     f"{self.num_elements} != {len(args_flattened)}"
     19 for arg, attr in zip(args_flattened, self._attrs):
     20     setattr(self, attr, arg)

AssertionError: 3 != 2
```

Reply:
Thanks for clarifying and the update. Is this an error that a user can make? If so, would it be more appropriate to provide a `ValueError` or `RuntimeError` with a message like `float32x3 expected 3 values, received 2`?